File: src/2532_Time_to_Cross_a_Bridge.py
# 2532. 过桥的时间
# https://leetcode.cn/problems/time-to-cross-a-bridge/description/


# 方法一：优先队列
# 思路与算法

# 在本题中，工人共有 4 种状态：

# 在左侧等待
# 在右侧等待
# 在左侧工作（放下所选箱子）
# 在右侧工作（搬起所选箱子）

# 每一种工作状态都有相应的优先级计算方法，因此我们用 4 个优先队列来存放处于每种状态下的工人集合：

# 在左侧等待的工人：wait_left，题目中定义的效率越高，优先级越高。
# 在右侧等待的工人：wait_right，题目中定义的效率越高，优先级越高。
# 在左侧工作的工人：work_left，完成时间越早，优先级越高。
# 在右侧工作的工人：work_right，完成时间越早，优先级越高。

# 我们令 remain 表示右侧还有多少个箱子需要搬运，当 remain>0 时，搬运工作需要继续。
# 除此之外，题目求解的是最后一个回到左边的工人的到达时间，
# 因此当右侧还有工人在等待或在工作时（即 work_right或 wait_right 不为空），搬运工作就需要继续：

# 若 work_left中的工人在此刻已经完成工作，则需要将它们取出并分别加入到 wait_left 和 wait_right 中。
# 若 wait_right 不为空，则取其中优先级最低的工人过桥，将其加入到 work_left 队列中，并将时间更改为过桥后的时间。继续下一轮循环。
# 若 remain>0，并且 wait_left 不为空，则需要取优先级最低的工人过桥去取箱子，将其加入到 work_right 队列中，
# 令 remain 减 1，并将时间更改为过桥后的时间。继续下一轮循环。
# 若 2 和 3 都不满足，表示当前没有人需要过桥，当前时间应该过渡到 work_left 和 work_right 中的最早完成时间。然后继续下一轮循环。


# 这是一个复杂模拟题，值得标Hard。“等待中”和“工作中”各需要2个优先队列（注意关键字有所不同），
# 每次循环开始时，先根据当前时间节点判断两侧是否有“工作中”的工人变为”等待中“，
# 然后再从”等待中“的工人里按规则选出过桥的那个，再将其加入桥另一侧的”工作中“。
# 如果没有”等待中“，那么当前时间就要快进到最近的会出现”等待中“的时间点。

from typing import List

import heapq

class Solution:
    def findCrossingTime(self, n: int, k: int, time: List[List[int]]) -> int:
        # 将每个工人的编号附加到 time 列表中，以便追踪
        for i in range(len(time)):
            time[i].append(i)
            
        # 根据工人往返桥的时间对 time 进行排序
        time.sort(key=lambda x: [-x[0]-x[2], -x[4]])

        current_time = 0
        left_side_waiting = []  # 在桥左侧等待的工人
        right_side_waiting = []  # 在桥右侧等待的工人
        left_side_heap = [[j, 0] for j in range(k)]  # 桥左侧的工人堆
        right_side_heap = []  # 桥右侧的工人堆
        heapq.heapify(left_side_heap)  # 将列表转换为堆
        boxes_left = n
        boxes_moved = 0

        while boxes_left:
            if right_side_heap:  # 如果右侧堆中有工人
                worker_idx, _ = heapq.heappop(right_side_heap)
                heapq.heappush(left_side_waiting, [current_time + time[worker_idx][2] + time[worker_idx][3], worker_idx])
                current_time += time[worker_idx][2]
                boxes_left -= 1
            elif left_side_heap and boxes_moved < n:  # 如果左侧堆中有工人并且尚未移动所有箱子
                worker_idx, _ = heapq.heappop(left_side_heap)
                heapq.heappush(right_side_waiting, [current_time + time[worker_idx][0] + time[worker_idx][1], worker_idx])
                current_time += time[worker_idx][0]
                boxes_moved += 1
            else:
                # 将当前时间前进到下一个重要时间点 - 当前左侧任务的结束或右侧任务的结束
                current_time = min(
                    left_side_waiting[0][0] if left_side_waiting and boxes_moved < n else 10**9, 
                    right_side_waiting[0][0] if right_side_waiting else 10**9
                )

            # 处理所有可以在当前时间完成的右侧工人
            while right_side_waiting and right_side_waiting[0][0] <= current_time:
                heapq.heappush(right_side_heap, [right_side_waiting[0][1], right_side_waiting[0][0]])
                heapq.heappop(right_side_waiting)
                
            # 处理所有可以在当前时间完成的左侧工人
            while left_side_waiting and left_side_waiting[0][0] <= current_time:
                heapq.heappush(left_side_heap, [left_side_waiting[0][1], left_side_waiting[0][0]])
                heapq.heappop(left_side_waiting)

        return current_time

File: src/2532_Time_to_Cross_a_Bridge.rs
// 2532_Time_to_Cross_a_Bridge
// https://leetcode.cn/problems/time-to-cross-a-bridge/

// There are k workers who want to move n boxes from an old warehouse to a new one. You are given the two integers n and k, and a 2D integer array time of size k x 4 where time[i] = [leftToRighti, pickOldi, rightToLefti, putNewi].

// The warehouses are separated by a river and connected by a bridge. The old warehouse is on the right bank of the river, and the new warehouse is on the left bank of the river. Initially, all k workers are waiting on the left side of the bridge. To move the boxes, the ith worker (0-indexed) can :

// Cross the bridge from the left bank (new warehouse) to the right bank (old warehouse) in leftToRighti minutes.
// Pick a box from the old warehouse and return to the bridge in pickOldi minutes. Different workers can pick up their boxes simultaneously.
// Cross the bridge from the right bank (old warehouse) to the left bank (new warehouse) in rightToLefti minutes.
// Put the box in the new warehouse and return to the bridge in putNewi minutes. Different workers can put their boxes simultaneously.

// A worker i is less efficient than a worker j if either condition is met:
// leftToRighti + rightToLefti > leftToRightj + rightToLeftj
// leftToRighti + rightToLefti == leftToRightj + rightToLeftj and i > j

// The following rules regulate the movement of the workers through the bridge :
// If a worker x reaches the bridge while another worker y is crossing the bridge, x waits at their side of the bridge.
// If the bridge is free, the worker waiting on the right side of the bridge gets to cross the bridge. If more than one worker is waiting on the right side, the one with the lowest efficiency crosses first.
// If the bridge is free and no worker is waiting on the right side, and at least one box remains at the old warehouse, the worker on the left side of the river gets to cross the bridge. If more than one worker is waiting on the left side, the one with the lowest efficiency crosses first.
// Return the instance of time at which the last worker reaches the left bank of the river after all n boxes have been put in the new warehouse.

// 1 <= n, k <= 104
// time.length == k
// time[i].length == 4
// 1 <= leftToRighti, pickOldi, rightToLefti, putNewi <= 1000

// We can use a priority queue to query over the least efficient worker.

use std::collections::BinaryHeap;
use std::cmp::Reverse;

impl Solution {
    pub fn find_crossing_time(n: i32, k: i32, time: Vec<Vec<i32>>) -> i32 {
        let mut time: Vec<_> = time.into_iter().enumerate().map(|(i, mut t)| {
            t.push(i as i32);
            t
        }).collect();
        time.sort_by_key(|x| (Reverse(x[0] + x[2]), Reverse(x[4])));

        let mut current_time = 0;
        let mut left_side_waiting = BinaryHeap::new();
        let mut right_side_waiting = BinaryHeap::new();
        let mut left_side_heap: BinaryHeap<_> = (0..k).map(|j| (Reverse(0), j)).collect();
        let mut right_side_heap = BinaryHeap::new();
        let mut boxes_left = n;
        let mut boxes_moved = 0;

        while boxes_left > 0 {
            if let Some((_, worker_idx)) = right_side_heap.pop() {
                left_side_waiting.push((Reverse(current_time + time[worker_idx as usize][2] + time[worker_idx as usize][3]), worker_idx));
                current_time += time[worker_idx as usize][2];
                boxes_left -= 1;
            } else if boxes_moved < n {
                let (_, worker_idx) = left_side_heap.pop().unwrap();
                right_side_waiting.push((Reverse(current_time + time[worker_idx as usize][0] + time[worker_idx as usize][1]), worker_idx));
                current_time += time[worker_idx as usize][0];
                boxes_moved += 1;
            } else {
                current_time = *[
                    left_side_waiting.peek().map(|&(Reverse(t), _)| t).unwrap_or(10_i32.pow(9)),
                    right_side_waiting.peek().map(|&(Reverse(t), _)| t).unwrap_or(10_i32.pow(9)),
                ].iter().min().unwrap();
            }

            while right_side_waiting.peek().map_or(false, |&(Reverse(t), _)| t <= current_time) {
                let (_, worker_idx) = right_side_waiting.pop().unwrap();
                right_side_heap.push((Reverse(worker_idx), current_time));
            }

            while left_side_waiting.peek().map_or(false, |&(Reverse(t), _)| t <= current_time) {
                let (_, worker_idx) = left_side_waiting.pop().unwrap();
                left_side_heap.push((Reverse(worker_idx), current_time));
            }
        }

        current_time
    }
}

// Struct std::cmp::ReverseCopy 

#[repr(transparent)]
pub struct Reverse<T>(pub T);
// 用于逆序排序的辅助结构体。

// 该结构是一个帮助器，可与 Vec::sort_by_key 等函数一起使用，并且可以用于对键的一部分进行反向排序。


// use std::cmp::Reverse;

// let mut v = vec![1, 2, 3, 4, 5, 6];
// v.sort_by_key(|&num| (num > 3, Reverse(num)));
// assert_eq!(v, vec![3, 2, 1, 6, 5, 4]);

/// Rust does not include a built-in priority queue, 
/// but it has a BinaryHeap type in the standard library which can be used as a max priority queue. 
/// If you want to use it as a min priority queue, you need to wrap your values into std::cmp::Reverse.

// Here is an example of using BinaryHeap:
use std::collections::BinaryHeap;

fn main() {
    let mut heap = BinaryHeap::new();

    heap.push(3);
    heap.push(5);
    heap.push(1);

    // Heap is a max heap, so the elements will be popped in descending order
    while let Some(val) = heap.pop() {
        println!("{}", val);
    }
}

/// In this code:

/// We create a BinaryHeap with BinaryHeap::new.
/// We push some values onto the heap with heap.push.
/// We pop all values from the heap with heap.pop in a loop, 
/// which returns Some(val) until the heap is empty, at which point it returns None and the loop stops.
/// This code will print out:

// 5
// 3
// 1

/// If you want to implement a min-heap (smallest value has the highest priority), you can use std::cmp::Reverse:

use std::collections::BinaryHeap;
use std::cmp::Reverse;

fn main() {
    let mut heap = BinaryHeap::new();

    heap.push(Reverse(3));
    heap.push(Reverse(5));
    heap.push(Reverse(1));

    // Heap is now a min heap because of the Reverse, so the elements will be popped in ascending order
    while let Some(Reverse(val)) = heap.pop() {
        println!("{}", val);
    }
}

// This code will print out:

// 1
// 3
// 5

/// In this code, Reverse is a newtype that changes the sort order by reversing the comparison results.

/// If you need more complex priority queue functionality (for example, 
/// a priority queue that supports updating the priority of items), 
/// you may need to use a third-party crate such as priority-queue or heapless.